clippy_lints/src/ptr_offset_by_literal.rs

```diff
@@ -0,0 +1,156 @@
+use clippy_utils::diagnostics::{span_lint, span_lint_and_sugg};
+use clippy_utils::source::SpanRangeExt;
+use clippy_utils::sym;
+use rustc_ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use std::cmp::Ordering;
+use std::fmt;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of the `offset` pointer method with an integer
+    /// literal.
+    ///
+    /// ### Why is this bad?
+    /// The `add` and `sub` methods more accurately express the intent.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let vec = vec![b'a', b'b', b'c'];
+    /// let ptr = vec.as_ptr();
+    ///
+    /// unsafe {
+    ///     ptr.offset(-8);
+    /// }
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```no_run
+    /// let vec = vec![b'a', b'b', b'c'];
+    /// let ptr = vec.as_ptr();
+    ///
+    /// unsafe {
+    ///     ptr.sub(8);
+    /// }
+    /// ```
+    #[clippy::version = "CURRENT_RUSTC_VERSION"]
+    pub PTR_OFFSET_BY_LITERAL,
+    complexity,
+    "unneeded pointer offset"
+}
+
+declare_lint_pass!(PtrOffsetByLiteral => [PTR_OFFSET_BY_LITERAL]);
+
+impl<'tcx> LateLintPass<'tcx> for PtrOffsetByLiteral {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        // Check if the expressions is a ptr.offset or ptr.wrapping_offset method call
+        let Some((receiver_expr, arg_expr, method)) = expr_as_ptr_offset_call(cx, expr) else {
+            return;
+        };
+
+        // Check if the argument to the method call is a (negated) literal.
+        let Some(literal) = expr_as_literal(arg_expr) else {
+            return;
+        };
+
+        let msg = format!("use of `{method}` with a literal");
+        if let Some(sugg) = build_suggestion(cx, method, receiver_expr, literal) {
+            span_lint_and_sugg(
+                cx,
+                PTR_OFFSET_BY_LITERAL,
+                expr.span,
+                msg,
+                "try",
+                sugg,
+                Applicability::MachineApplicable,
+            );
+        } else {
+            span_lint(cx, PTR_OFFSET_BY_LITERAL, expr.span, msg);
+        }
+    }
+}
+
+fn get_literal_bits<'tcx>(expr: &'tcx Expr<'tcx>) -> Option<u128> {
+    let ExprKind::Lit(lit) = expr.kind else {
```

Comment:
nit: a let-chain is a bit more concise
```suggestion
    if let ExprKind::Lit(lit) = expr.kind
        && let LitKind::Int(packed_u128, _) = lit.node
    {
        Some(packed_u128.get())
    } else {
        None
    }
```
(the suggestion seems to be formatted weirdly, but it should (?) work correctly)